system_modules/napaudio/src/audio/node/levelmeternode.cpp

```diff
@@ -61,43 +52,55 @@ namespace nap
 
 			if (inputBuffer == nullptr)
 				return;
-
-			for (auto& sample : *inputBuffer) {
-				mBuffer[mIndex++] = sample;
-				if (mIndex == mBuffer.size()) {
-					mIndex = 0;
-					mDirty.set();
-				}
+            
+			switch (mType)
+			{
+				case PEAK:
+					for (auto& sample : *inputBuffer)
+					{
+						// keep track of peak, sample by sample
+						float newValue = sample;
+						if (newValue < 0.f)
+							newValue = -newValue;
+						if (newValue > mPeakTemp)
+							mPeakTemp = newValue;
+						
+						// reset temp and update the output value once every 'window'
+						mIndex++;
+						if (mIndex == mWindowSizeInSamples)
+						{
+							mPeak = mPeakTemp;
+							mPeakTemp = 0.f;
+							mIndex = 0;
+						}
+					}
+					
+					mValue.store(mPeak);
+					break;
+
+				default: // RMS
+					for (auto& sample : *inputBuffer)
+					{
+						// updated squared sum
```

Comment:
Why is there still a mSquaredBuffer? You can just update the sum every sample without having to store the samples in a buffer.

Reply:
> Why is there still a mSquaredBuffer? You can just update the sum every sample without having to store the squared samples in a buffer.

Hm, how would you calculate the sum of the last X squared samples without storing the last X squared samples?

To explain the implemented algorithm: mSquaredSum is the sum of the last X squared samples (where X depends on the "analysis window size"). This sum is updated each sample by subtracting the value of the oldest sample and adding the value of the latest sample.

Reply:
Ahh I overlooked the subtraction of the oldest value in the buffer. :)
Ok then I get it, all good!